firedrake/slate/slac/kernel_builder.py

```diff
@@ -379,3 +391,418 @@ def needs_mesh_layers(self):
+        In particular needed for the right shape of scalar tensors.
+        """
+        if tensor.shape == ():
+            return (1, )  # scalar tensor
+        else:
+            return tensor.shape
+
+    def _setup(self):
+        """A setup method to initialize all the local assembly
+        kernels generated by TSFC. This function also collects any
+        information regarding orientations and extra include directories.
+        """
+        include_dirs = []
+        templated_subkernels = []
+
+        assembly_calls = OrderedDict([(it, []) for it in self.supported_integral_types])
+        coords = None
+        needs_cell_orientations = False
+        needs_cell_sizes = False
+        needs_cell_facets = False
+        needs_mesh_layers = False
+        num_facets = 0
+
+        # For all terminal tensors provided by TSFC
+        for pos, cxt_kernel in enumerate(self.context_kernels):
+            coefficients = cxt_kernel.coefficients
+            integral_type = cxt_kernel.original_integral_type
+            tensor = cxt_kernel.tensor
+            temp = self.temps[tensor]
+            mesh = tensor.ufl_domain()
+
+            if integral_type not in self.supported_integral_types:
+                raise ValueError("Integral type '%s' not recognized" % integral_type)
+
+            # Explicit checking of coordinates
+            coordinates = cxt_kernel.tensor.ufl_domain().coordinates
+            if coords is not None:
+                assert coordinates == coords, "Mismatching coordinates!"
+            else:
+                coords = coordinates
+
+            for inner in cxt_kernel.tsfc_kernels:
+                kinfo = inner.kinfo
+                reads = []
+                inames = []
+
+                # Populate subkernel call to tsfc
+                templated_subkernels.append(kinfo.kernel.code)
+                include_dirs.extend(kinfo.kernel._include_dirs)
+
+                # Generation of output variable of loopy kernel
+                if tensor.is_mixed:
+                    # For the mixed case, the output is a slice of the matrix/vector
+                    block_index = inner.indices
+                    extent = ()
+                    offset = ()
+                    slices = ()
+
+                    # e.g. for rank 2 mixed tensor you get two extents, offsets
+                    for i, j in enumerate(block_index):
+                        extent += (tensor.shapes[i][j],)
+                        offset += (sum(tensor.shapes[i][:j]),)
+                    idx = self.create_index(extent, "block_idx"+kinfo.kernel.name)
+
+                    # Indices for subscripting the tensor
+                    for i, j in enumerate(block_index):
+                        slices += (pym.Sum((offset[i], idx[i])),)
+
+                    output = SubArrayRef(idx,
+                                         pym.Subscript(pym.Variable(self.gem_loopy_dict[temp].name), slices))
```

Comment:
I think I don't understand what is going on here. You could just build a single tsfc kernel that assembles the mixed terminal tensor, but maybe that is not wanted.

Reply:
I think the logic doing all the mapping between gem and loopy objects and so forth would be easier if you first translate slate->gem, in doing so you can remember gem Variables and the slate tensors they come from.
Now you can traverse these variables and generate these callables for each of them. Rather than calling tsfc compilation much earlier and so forth. When you do this stage you have to collect information about the arguments for each subkernel (as a mapping from the `expression.coefficients()`). Then the gluing is, I think, easier.

Reply:
> I think I don't understand what is going on here. You could just build a single tsfc kernel that assembles the mixed terminal tensor, but maybe that is not wanted.

I did not know that this is possible. The current Tsfc_driver in slac does not specify split in `compile_terminal_form`, is it right that if if specify `split=false`, I don't get the blocks in separate kernels anymore, but in one?

Reply:
Yes, if you say `split=False` you get one kernel for a mixed form. It might be that this is not desired though.

Reply:
Can you have different integral_types in the blocks of one matrix?

Reply:
Yes, but not in a single kernel. e.g. `u*v*dx + u*v*ds` is different integral_types in the same block.

Reply:
If we request the kernel from Tsfc to be not split, do the coefficients that the kernel reads still need to be split?

Reply:
Yes. You can avoid splitting of the coefficients by providing a set of `dont_split` coefficients. However, if you do that, then you have to change the interface in the global assembly loop inside `assemble` (since that right now always splits coefficients on sending them into the par_loop): the par_loop call has to agree with the API the kernel offers.

Reply:
That is, TSFC can handle mixed arguments as well as mixed coefficients, but depending on what you want to do, you may wanna give a split form (or split coefficients) to TSFC.  Which is what often happens from the Firedrake side.